### PR TITLE
Text preprocessing implementation

### PR DESCRIPTION
**Pull request by Louquinze**
feature_validator.py: added string option

meta_features.py: add string option

feature_type_text.py + text_encoding implements the new feature

*.csv and *.json files adapted to new feature type

### CODE REVIEW

File: autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json [hunk 1]
```diff
@@ -1,1197 +1 @@
...
..."3.125": 3, "6.25": 4, "12.5": 5, "25": 6, "50": 7, "100": 8}, {"0.390625": 0, "0.78125": 1, "1.5625": 2, "3.125": 3, "6.25": 4, "12.5": 5, "25": 6, "50": 7, "100": 8}], "config_to_task": {"b8184c0": [340], "05327ff": [2119], "73fc45b": [2122], "d38f297": [2123], "d033742": [3047], "8c75f1c": [3049], "074da43": [75089], "a479a3d": [75098], "d54bc21": [75148], "8cd6304": [75157], "b86e0b2": [75159], "978b964": [75171], "e249036": [126030], "f187687": [146574], "2f6c8d6": [146578], "a3ba205": [146597], "d618703": [146602], "c437bea": [166866], "77764f9": [166957], "05690ed": [167085], "c2b918a": [167088], "eb6f7f1": [167089], "c29f8bf": [167099], "2d4a38e": [167101], "7cab112": [189779], "68e6606": [189836], "811cba1": [189843], "fe0e803": [189863], "c832fa9": [189881], "a2734c9": [189882], "564103f": [189884], "07bebab": [189893]}, "input_directory": "/work/dlclarge2/feurerm-askl2paper/askl2paper/2020_IEEE_Autosklearn_experiments/experiment_scripts/60MIN/ASKL_getportfolio/RF_SH-eta4-i_holdout_iterative_es_if", "fidelities": "SH", "portfolio_size": 32, "seed": 0, "max_runtime": null, "start_time": 1622279454.1839142, "end_time": 1622279506.850947, "wallclock_time": 52.66703271865845}
```
Comment (mfeurer) [inline autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json:1]: Could you please keep the formatting intact (should be achieved with indent = 4 when writing the json)?

File: autosklearn/data/feature_validator.py [hunk 2]
```diff
@@ -25,10 +25,8 @@
 class FeatureValidator(BaseEstimator):
     """
     Checks the input data to Auto-Sklearn.
-
```
Comment (mfeurer) [inline autosklearn/data/feature_validator.py:28]: Please do not make such formatting changes. These will make it hard to merge later on.

File: autosklearn/data/feature_validator.py [hunk 3]
```diff
@@ -363,7 +357,7 @@ def list_to_dataframe(
             # Warn the user about dtypes or request him to use a dataframe
             for col in X_train.columns:
                 if X_train[col].dtype.name == 'string':
-                    X_train[col] = X_train[col].astype('category')
+                    X_train[col] = X_train[col].astype('string')
```
Comment (mfeurer) [inline autosklearn/data/feature_validator.py:360]: Do we actually have to change something here or can we just drop the whole for-loop?

File: autosklearn/data/feature_validator.py [hunk 4]
```diff
@@ -395,4 +389,4 @@ def list_to_dataframe(
                     self.logger.warning(f"Failed to format column {col} as {self.dtypes[col]}: {e}")
                     self.dtypes[col] = X_test[col].dtype.name.lower()
 
-        return X_train, X_test
+        return X_train, X_test
```
Comment (mfeurer) [inline autosklearn/data/feature_validator.py:392]: Please undo the formatting change here.

File: autosklearn/data/feature_validator.py [hunk 5]
```diff
@@ -129,9 +125,10 @@ def fit(
                     ))
 
                 for ft in self.feat_type.values():
-                    if ft.lower() not in ['categorical', 'numerical']:
-                        raise ValueError('Only `Categorical` and `Numerical` are '
-                                         'valid feature types, you passed `%s`' % ft)
+                    print(ft)
```
Comment (mfeurer) [inline autosklearn/data/feature_validator.py:128]: No print statements please

File: autosklearn/metalearning/metalearning/meta_base.py [hunk 6]
```diff
@@ -46,6 +46,8 @@ def __init__(self, configuration_space, aslib_directory, logger):
         for algorithm_id in self.configurations:
             configuration = self.configurations[algorithm_id]
             try:
+                # ToDo check if all stored configurations are actually readable
```
Comment (mfeurer) [inline autosklearn/metalearning/metalearning/meta_base.py:49]: What does this mean?

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py [hunk 7]
```diff
@@ -29,22 +31,22 @@
...
-        pipeline: Optional[BasePipeline] = None,
-        dataset_properties: Optional[DATASET_PROPERTIES_TYPE] = None,
-        include: Optional[Dict[str, str]] = None,
-        exclude: Optional[Dict[str, str]] = None,
-        random_state: Optional[Union[int, np.random.RandomState]] = None,
-        init_params: Optional[Dict[str, Any]] = None,
-        feat_type: Optional[Dict[Union[str, int], str]] = None,
-        force_sparse_output: bool = False,
-        column_transformer: Optional[sklearn.compose.ColumnTransformer] = None,
+            self,
```
Comment (mfeurer) [inline autosklearn/pipeline/components/data_preprocessing/feature_type.py:39]: Please do not make such formatting changes. These make it hard for me to review your PR.

File: test/test_automl/test_automl.py [hunk 8]
```diff
@@ -641,6 +640,8 @@ def test_load_best_individual_model(metric, dask_client):
         assert automl.score(X_test, Y_test) > 0.9
     elif metric.name == 'log_loss':
         # Seen values in github actions of 0.6978304740364537
+        # ToDo error failed test, just bad prediction ... 1.9 is worse than 0.7
```
Comment (mfeurer) [inline test/test_automl/test_automl.py:643]: This appears to be unrelated to this PR and actually, this test doesn't fail right now. Could you please undo the changes?

File: test/test_pipeline/test_regression.py [hunk 9]
```diff
@@ -284,7 +284,8 @@ def test_get_hyperparameter_search_space(self):
         conditions = cs.get_conditions()
         hyperparameters = cs.get_hyperparameters()
         forbiddens = cs.get_forbiddens()
-        self.assertEqual(156, len(hyperparameters))
+        # 12 new hps 156 --> 168
```
Comment (mfeurer) [inline test/test_pipeline/test_regression.py:287]: Please drop such comments, they'll be outdated in the future.

File: autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json (hunk 1 above)
Comment (eddiebergman) [reply autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json:1]: While this is a topic, should we sort keys as well where possible?
```python
with open(...) as file:
    json.dump(dictionary, file, indent=4, sort_keys=True)
```

File: autosklearn/data/feature_validator.py (hunk 3 above)
Comment (eddiebergman) [reply autosklearn/data/feature_validator.py:360]: This for-loop does seem unnecessary, if `dtype.name == 'string'` then there's no need to convert it.

File: autosklearn/pipeline/components/data_preprocessing/text_encoding/bag_of_word_encoding.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+from sklearn.feature_extraction.text import CountVectorizer
+
+
+class BagOfWordEncoder(AutoSklearnPreprocessingAlgorithm):
+    def __init__(
+            self,
+            min_df_choice=None,
+            min_df_absolute=None,
+            min_df_relative=None,
+            random_state: Optional[Union[int, np.random.RandomState]] = None
```
Comment (eddiebergman) [inline autosklearn/pipeline/components/data_preprocessing/text_encoding/bag_of_word_encoding.py:23]: Add types the parameters if you can

**Review by eddiebergman — COMMENTED** [review]
Did a first pass review with some minor revisions, main thing is to fix the indentation from the parameters in constructors and importantly to include the `random.state`, it took a week to try fix somethings to be more detirministic in the past and they easily slip by.

Looks okay for now, I'll do another review over it soon but looks good otherwise! :)

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py [hunk 11]
```diff
@@ -112,23 +130,39 @@ def fit(self, X: SUPPORTED_FEAT_TYPES, y: Optional[SUPPORTED_TARGET_TYPES] = Non
...
-        # For the other cases, both transformers are used
-        else:
-            sklearn_transf_spec = [
-                ("categorical_transformer", self.categ_ppl, categorical_features),
-                ("numerical_transformer", self.numer_ppl, numerical_features)
-            ]
+        n, c, t = len(numerical_features) != 0, len(categorical_features) != 0, len(
+            text_features) != 0
+        n, c, t = str(int(n)), str(int(c)), str(int(t))
+        sklearn_transf_spec = features_dictionary[t + c + n]
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/feature_type.py:165]: On second pass, here's a condensed version that doesn't require the dictionary and is more extendable in the future:

```python
spec = [
    (name, transformer, feature_columns)
    for name, transformer, feature_columns
    in [
        ("text_transformer", self.txt_ppl, text_features),
        ("categorical_transformer", self.categ_ppl, categorical_features),
        ("numerical_transformer", self.numer_ppl, numerical_features)
    ]
    if len(feature_columns) > 0
]
```

File: autosklearn/metalearning/metalearning/meta_base.py (hunk 6 above)
Comment (Louquinze) [reply autosklearn/metalearning/metalearning/meta_base.py:49]: I have noticed that some configurations cant be loaded because in the referring configurations.csv the value 999 instead of an float < 1.0 is stored i think that is the bug which makes waringings like:

[WARNING] [2021-11-04 10:16:34,533:Client-AutoMLSMBO(1)::e6ac0eb2-3d4f-11ec-9631-98bb1e1dd0b6] Configuration 555 not found

if you want to reconstruct this the problem i found you can comment out the try / except statement and just add the line where the configurations loaded without try statement then you will get an error like:

  File "/home/lukas/Python_Projects/NewBOWDev/auto-sklearn/autosklearn/metalearning/metalearning/meta_base.py", line 52, in __init__
    (Configuration(configuration_space, values=configuration))
  File "ConfigSpace/configuration_space.pyx", line 1412, in ConfigSpace.configuration_space.Configuration.__init__
ValueError: Trying to set illegal value '999' (type '<class 'int'>') for hyperparameter 'classifier:mlp:beta_2, Type: Constant, Value: 0.999' (default-value has type '<class 'float'>').

File: autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+class FeatureReduction(AutoSklearnPreprocessingAlgorithm):
+    """
+    Reduces the features created by a bag of words encoding
+    """
+
+    def __init__(
+            self,
+            n_components=None,
+            random_state: Optional[Union[int, np.random.RandomState]] = None
```
[the remark this answers is not in the document]
Comment (Louquinze) [reply autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py:24]: what should i do instead ?

File: autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    ) -> None:
+        self.n_components = n_components
+        self.random_state = random_state
+
+    def fit(self, X: PIPELINE_DATA_DTYPE, y: Optional[PIPELINE_DATA_DTYPE] = None
+            ) -> 'FeatureReduction':
+        X = csr_matrix(X)
```
[the remark this answers is not in the document]
Comment (Louquinze) [reply autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py:31]: i do not remember what i excellently thought about, but i think it was because in the truncedsvd documentation is mentioned that one of the main advantages of  truncedsvd in comparison to pca is that it can deal better with sparse matrices but i think you are right and it does not make so much sense here.

File: autosklearn/metalearning/metalearning/meta_base.py (hunk 6 above)
Comment (eddiebergman) [reply autosklearn/metalearning/metalearning/meta_base.py:49]: That's an interesting bug to have discovered, we should probably make this as an issue as it's come up a few times in the past.

File: autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py:31]: Hmm, that is true but I'm not sure it requires an explicit sparse matrix type, more generally that the method itself works on matrices with sparse entries. I would hesitate to convert the matrix to sparse unless we can guarantee that it is actually sparse before converting it.

@mfeurer thoughts?

File: autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/pipeline/components/data_preprocessing/feature_reduction/truncated_svd.py:24]: Sorry, I just meant to say that you can add a type to `n_components` where the type is `Optional[int]`

File: autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json (hunk 1 above)
Comment (mfeurer) [reply autosklearn/experimental/balanced_accuracy/askl2_portfolios/RF_SH-eta4-i_holdout_iterative_es_if.json]: We would have to do so in a different repo.

File: autosklearn/metalearning/metalearning/meta_base.py (hunk 6 above)
Comment (mfeurer) [reply autosklearn/metalearning/metalearning/meta_base.py]: I cannot reproduce this.
